File: api/app/infrastructure/external/browser/playwright_browser.py
import asyncio
import logging
from typing import TYPE_CHECKING, Any, List, Optional

from markdownify import markdownify
from playwright.async_api import async_playwright

from app.domain.external.browser import Browser as BrowserProtocol
from app.domain.external.llm import LLM
from app.domain.models.tool_result import ToolResult
from app.infrastructure.external.browser.playwrightBrowserFunc import (
    GET_INTERACTIVE_VISIBLE_CONTENT_FUNC,
    GET_VISIBLE_CONTENT_FUNC,
    INJECT_CONSOLE_FUNC,
)

if TYPE_CHECKING:
    from playwright.async_api import Browser as PwBrowser, Page, Playwright

logger = logging.getLogger(__name__)


class PlayWrightBrowser(BrowserProtocol):
    def __init__(self, cdp_url: str, llm: Optional[LLM] = None):
        self.cdp_url = cdp_url
        self.llm: Optional[LLM] = llm
        self.playwright: Optional["Playwright"] = None
        self.browser: Optional["PwBrowser"] = None
        self.page: Optional["Page"] = None

    async def initialize(self) -> bool:
        max_retries = 5
        retry_interval = 1

        for attempt in range(max_retries):
            try:
                self.playwright = await async_playwright().start()
                logger.info("Connecting to CDP: %s", self.cdp_url)
                self.browser = await self.playwright.chromium.connect_over_cdp(
                    self.cdp_url,
                    is_local=self.cdp_url.startswith("http://127.0.0.1")
                    or self.cdp_url.startswith("http://localhost"),
                )

                contexts = self.browser.contexts
                if contexts and contexts[0].pages:
                    self.page = contexts[0].pages[-1]
                else:
                    context = (
                        contexts[0] if contexts else await self.browser.new_context()
                    )
                    self.page = await context.new_page()
                return True
            except Exception as e:
                await self.cleanup()
                logger.error(f"初始化浏览器失败: {e}")

                if attempt == max_retries - 1:
                    logger.error(f"初始化浏览器失败,已重试({max_retries})次")
                    return False

                await asyncio.sleep(retry_interval)
                retry_interval = min(retry_interval * 2, 10)

    async def cleanup(self) -> None:
        try:
            if self.browser:
                await self.browser.close()
            elif self.page and not self.page.is_closed():
                await self.page.close()

            if self.playwright:
                await self.playwright.stop()
        except Exception as e:
            logger.error("清理浏览器失败: %s", e)
        finally:
            self.page = None
            self.browser = None
            self.playwright = None

    async def _ensure_browser_exist(self) -> None:
        if not self.browser or not self.page:
            if not await self.initialize():
                raise Exception("初始化浏览器失败")

    async def _ensure_page_exist(self) -> None:
        await self._ensure_browser_exist()
        if not self.page:
            self.page = await self.browser.new_page()
        else:
            contexts = self.browser.contexts

            if contexts:
                default_context = contexts[0]
                pages = default_context.pages

                if pages:
                    latest_page = pages[-1]
                    if self.page != latest_page:
                        self.page = latest_page

    async def wait_for_page_load(self, time: int = 15) -> bool:
        await self._ensure_page_exist()

        start_time = asyncio.get_event_loop().time()
        check_interval = 5

        while asyncio.get_event_loop().time() - start_time < time:
            try:
                if await self.page.evaluate("document.readyState === 'complete'"):
                    return True
            except Exception as e:
                logger.error(f"等待页面加载失败: {e}")
                return False
            await asyncio.sleep(check_interval)
        return False

    async def _extract_content(self) -> str:
        visible_content = await self.page.evaluate(GET_VISIBLE_CONTENT_FUNC)
        markdown_content = markdownify(visible_content)

        MAX_CONTENT_LEN = min(len(markdown_content), 50000)

        if self.llm:
            response = await self.llm.invoke(
                messages=[
                    {
                        "role": "system",
                        "content": "你是一个网页内容提取工具，请提取网页中的所有信息，并以markdown格式返回。",
                    },
                    {
                        "role": "user",
                        "content": f"请提取网页中的所有信息，并以markdown格式返回。\n\n{markdown_content[:MAX_CONTENT_LEN]}",
                    },
                ]
            )
            return response.get("content", "")
        else:
            return markdown_content[:MAX_CONTENT_LEN]

    async def _extract_interactive_content(self) -> List[str]:
        logger.info(f"开始提取交互元素")
        await self._ensure_page_exist()

        self.page.interactive_elements_cache = []

        interactive_elements = await self.page.evaluate(
            GET_INTERACTIVE_VISIBLE_CONTENT_FUNC
        )
        logger.info(f"提取交互元素结束: {interactive_elements}")
        self.page.interactive_elements_cache = interactive_elements

        formatted_elements = []

        for element in interactive_elements:
            logger.info("formatting element: %s", element)
            formatted_elements.append(f"{element['index']}:{element['outerHTML']}")

        return formatted_elements

    async def navigate(self, url: str) -> ToolResult:
        await self._ensure_page_exist()
        try:
            logger.info(f"开始跳转: {url}")
            await self.page.goto(url)
            logger.info(f"跳转结束: {url}")
            data = await self._extract_interactive_content()
            logger.info(f"navigate data: {data}")
            return ToolResult(
                success=True,
                data=data,
                message="页面导航成功",
            )
        except Exception as e:
            logger.error(f"页面跳转失败: {e}")
            return ToolResult(success=False, message=f"页面导航失败: {e}")

    async def view_page(self) -> ToolResult:
        await self._ensure_page_exist()
        await self.wait_for_page_load()
        interactive_elements = await self._extract_interactive_content()

        return ToolResult(
            success=True,
            message="提取元素成功",
            data={
                "content": await self._extract_content(),
                "interactive_elements": interactive_elements,
            },
        )

    async def restart(self, url: str) -> ToolResult:
        await self.cleanup()
        return await self.navigate(url)

    async def scroll_up(self, to_top: Optional[bool] = None) -> ToolResult:
        await self._ensure_page_exist()
        try:
            if to_top:
                await self.page.evaluate("window.scrollTo(0, 0)")
            else:
                await self.page.evaluate("window.scrollBy(0, -window.innerHeight)")
            return ToolResult(success=True, message="滚动成功")
        except Exception as e:
            logger.error(f"滚动失败: {e}")
            return ToolResult(success=False, message=f"滚动失败: {e}")

    async def scroll_down(self, to_bottom: Optional[bool] = None) -> ToolResult:
        await self._ensure_page_exist()
        try:
            if to_bottom:
                await self.page.evaluate(
                    "window.scrollTo(0, document.body.scrollHeight)"
                )
            else:
                await self.page.evaluate("window.scrollBy(0, window.innerHeight)")
            return ToolResult(success=True, message="滚动成功")
        except Exception as e:
            logger.error(f"滚动失败: {e}")
            return ToolResult(success=False, message=f"滚动失败: {e}")

    async def screenshot(self, full_page: Optional[bool] = None) ->  bytes:
        await self._ensure_page_exist()
        if full_page:
            screenshot = await self.page.screenshot(full_page=True, type="png")
        else:
            screenshot = await self.page.screenshot(type="png")
        return screenshot

    async def console_exec(self, javascript: str) -> ToolResult:
        await self._ensure_page_exist()
        try:
            await self.page.evaluate(INJECT_CONSOLE_FUNC)
        except Exception as e:
            logger.error(f"注入window.console.logs失败: {str(e)}")

        result = await self.page.evaluate(javascript)
        return ToolResult(success=True, message="执行成功", data=result)

    async def console_view(self, max_lines: Optional[int] = None) -> ToolResult:
        await self._ensure_page_exist()

        logs = await self.page.evaluate(
            """
            () => {
                return window.console.logs || []
            }
        """
        )

        if max_lines is not None:
            logs = logs[-max_lines:]

        return ToolResult(success=True, message="获取成功", data=logs)

    async def _get_element_by_id(self, index: int) -> Optional[Any]:
        await self._ensure_page_exist()
        try:
            if (
                not hasattr(self.page, "interactive_elements_cache")
                or not self.page.interactive_elements_cache
                or index >= len(self.page.interactive_elements)
            ):
                return None
            selector = f'[data-manus-id="manus-element-{index}"]'
            return await self.page.query_selector(selector)
        except Exception as e:
            logger.error(f"获取元素失败: {e}")
            return None

    async def click(
        self,
        index: Optional[int] = None,
        coordinate_x: Optional[float] = None,
        coordinate_y: Optional[float] = None,
    ) -> ToolResult:
        await self._ensure_page_exist()
        try:
            if index is not None:
                element = await self._get_element_by_id(index)
                if element is None:
                    return ToolResult(success=False, message="元素不存在")
                is_visible = await self.page.evaluate(
                    """
                    (element) => {
                        if(!element) return false;
                        const rect = element.getBoundingClientRect();
                        const style = element.getComputedStyle();
                        return !(
                            rect.width === 0 ||
                            rect.height === 0 ||
                            style.display === 'none' ||
                            style.visibility === 'hidden' ||
                            style.opacity === '0'
                        );
                    }
                """,
                    element,
                )

                if not is_visible:
                    await self.page.evaluate(
                        """
                    (element) => {
                        element.scrollIntoView({
                            behavior: "smooth",
                            block: "center",
                        });
                    }
                    """,
                        element,
                    )
                await asyncio.sleep(1)
                await element.click(timeout=5000)
            elif coordinate_x is not None and coordinate_y is not None:
                await self.page.mouse.click(x=coordinate_x, y=coordinate_y)
            return ToolResult(success=True, message="点击成功")
        except Exception as e:
            logger.error(f"点击失败: {e}")
            return ToolResult(success=False, message=f"点击失败: {e}")

    async def input(
        self,
        index: Optional[int] = None,
        coordinate_x: Optional[float] = None,
        coordinate_y: Optional[float] = None,
        text: Optional[str] = None,
        press_enter: Optional[bool] = None,
    ) -> ToolResult:
        await self._ensure_page_exist()
        if index is not None:
            try:
                element = await self._get_element_by_id(index)
                if element is None:
                    return ToolResult(success=False, message="元素不存在")
                try:
                    await element.fill("")
                    await element.type(text)
                except Exception:
                    await element.click()
                    await element.type(text)
            except Exception as e:
                logger.error(f"输入失败: {e}")
                return ToolResult(success=False, message=f"输入失败: {e}")
        elif coordinate_x is not None and coordinate_y is not None:
            try:
                await self.page.mouse.click(x=coordinate_x, y=coordinate_y)
                await self.page.keyboard.type(text)
            except Exception as e:
                logger.error(f"输入失败: {e}")
                return ToolResult(success=False, message=f"输入失败: {e}")
        if press_enter:
            await self.page.keyboard.press("Enter")
        return ToolResult(success=True, message="输入成功")

    async def move_mouse(
        self,
        coordinate_x: Optional[float] = None,
        coordinate_y: Optional[float] = None,
    ) -> ToolResult:
        await self._ensure_page_exist()
        try:
            if coordinate_x is not None and coordinate_y is not None:
                await self.page.mouse.move(x=coordinate_x, y=coordinate_y)
            return ToolResult(success=True, message="移动成功")
        except Exception as e:
            logger.error(f"移动失败: {e}")
            return ToolResult(success=False, message=f"移动失败: {e}")

    async def press_key(self, key: str) -> ToolResult:
        await self._ensure_page_exist()
        try:
            await self.page.keyboard.press(key)
            return ToolResult(success=True, message="按键成功")
        except Exception as e:
            logger.error(f"按键失败: {e}")
            return ToolResult(success=False, message=f"按键失败: {e}")

    async def select_option(self, index: int, option: int) -> ToolResult:
        await self._ensure_page_exist()
        try:
            element = await self._get_element_by_id(index)
            if element is None:
                return ToolResult(success=False, message="元素不存在")
            await element.select_option(index=option)
            return ToolResult(success=True, message="选择成功")
        except Exception as e:
            logger.error(f"选择失败: {e}")
            return ToolResult(success=False, message=f"选择失败: {e}")

File: api-node/src/infrastructure/browser/playwright-browser.ts
import TurndownService from 'turndown';
import { chromium, Browser, Page, BrowserContext, ElementHandle } from 'playwright';
import { OpenAILLM } from '../llm/openai-llm';
import { toolSuccess, toolFailure } from '../../domain/tool-result';
import { logger } from '../../utils/logger';
import {
  GET_CONSOLE_LOGS_FUNC,
  GET_INTERACTIVE_VISIBLE_CONTENT_FUNC,
  GET_VISIBLE_CONTENT_FUNC,
  INJECT_CONSOLE_FUNC,
  IS_ELEMENT_VISIBLE_FUNC,
  SCROLL_ELEMENT_INTO_VIEW_FUNC,
} from './playwright-browser-func';

type InteractiveElement = {
  index: number;
  tagName: string;
  text: string;
  attributes: string;
  outerHTML: string;
};

/** Playwright 字符串 evaluate 会把入参当表达式执行；箭头函数字符串不调用时返回不可序列化的 Function → undefined */
function evaluatePageScript<T>(page: Page, script: string): Promise<T> {
  return page.evaluate(`(${script})()`);
}

function evaluatePageScriptWithArg<T>(page: Page, script: string, arg: unknown): Promise<T> {
  return page.evaluate(`(${script})`, arg);
}

export class PlaywrightBrowser {
  private browser: Browser | null = null;
  private page: Page | null = null;
  private interactiveCache: InteractiveElement[] = [];
  private readonly turndown = new TurndownService();

  constructor(
    private readonly cdpUrl: string,
    private readonly llm?: OpenAILLM,
  ) {}

  async initialize(): Promise<boolean> {
    let retryInterval = 1000;
    for (let attempt = 0; attempt < 5; attempt++) {
      try {
        logger.info(`Connecting to CDP: ${this.cdpUrl}`);
        this.browser = await chromium.connectOverCDP(this.cdpUrl);
        await this.ensureLatestPage();
        return true;
      } catch (e) {
        logger.error(`初始化浏览器失败: ${(e as Error).message}`);
        await this.cleanup();
        if (attempt === 4) return false;
        await new Promise((r) => setTimeout(r, retryInterval));
        retryInterval = Math.min(retryInterval * 2, 10_000);
      }
    }
    return false;
  }

  async cleanup(): Promise<void> {
    try {
      if (this.browser) await this.browser.close();
    } catch {
      /* ignore */
    }
    this.browser = null;
    this.page = null;
    this.interactiveCache = [];
  }

  private async ensureLatestPage(): Promise<Page> {
    if (!this.browser) {
      const ok = await this.initialize();
      if (!ok || !this.browser) throw new Error('初始化浏览器失败');
    }
    const contexts = this.browser!.contexts();
    if (contexts.length > 0 && contexts[0].pages().length > 0) {
      this.page = contexts[0].pages()[contexts[0].pages().length - 1];
    } else {
      const ctx: BrowserContext =
        contexts.length > 0 ? contexts[0] : await this.browser!.newContext();
      this.page = await ctx.newPage();
    }
    return this.page!;
  }

  private async waitForPageLoad(timeoutSec = 15): Promise<boolean> {
    const page = await this.ensureLatestPage();
    const deadline = Date.now() + timeoutSec * 1000;
    while (Date.now() < deadline) {
      try {
        const ready = await page.evaluate('document.readyState === "complete"');
        if (ready) return true;
      } catch (e) {
        logger.error(`等待页面加载失败: ${(e as Error).message}`);
        return false;
      }
      await new Promise((r) => setTimeout(r, 500));
    }
    return false;
  }

  private async extractContent(): Promise<string> {
    const page = await this.ensureLatestPage();
    const html = await evaluatePageScript<string>(page, GET_VISIBLE_CONTENT_FUNC);
    let markdown = this.turndown.turndown(String(html));
    const maxLen = Math.min(markdown.length, 50_000);
    markdown = markdown.slice(0, maxLen);

    if (this.llm) {
      try {
        const message = await this.llm.invoke([
          {
            role: 'system',
            content: '你是一个网页内容提取工具，请提取网页中的所有信息，并以markdown格式返回。',
          },
          {
            role: 'user',
            content: `请提取网页中的所有信息，并以markdown格式返回。\n\n${markdown}`,
          },
        ]);
        return message.content ?? markdown;
      } catch (e) {
        logger.warn(`LLM 内容提取失败，使用 Turndown 结果: ${(e as Error).message}`);
      }
    }
    return markdown;
  }

  private async extractInteractiveContent(): Promise<string[]> {
    const page = await this.ensureLatestPage();
    const elements =
      (await evaluatePageScript<InteractiveElement[]>(page, GET_INTERACTIVE_VISIBLE_CONTENT_FUNC)) ?? [];
    this.interactiveCache = elements;
    return elements.map((el) => `${el.index}:${el.outerHTML}`);
  }

  private async getElementByIndex(index: number): Promise<ElementHandle | null> {
    const page = await this.ensureLatestPage();
    if (index < 1 || index > this.interactiveCache.length) {
      return null;
    }
    return page.$(`[data-manus-id="manus-element-${index}"]`);
  }

  async viewPage() {
    try {
      await this.waitForPageLoad();
      const page = await this.ensureLatestPage();
      const interactiveElements = await this.extractInteractiveContent();
      const content = await this.extractContent();
      return toolSuccess({ content, interactive_elements: interactiveElements, url: page.url() }, '提取元素成功');
    } catch (e) {
      return toolFailure(`查看页面失败: ${(e as Error).message}`);
    }
  }

  async navigate(url: string) {
    try {
      const page = await this.ensureLatestPage();
      logger.info(`开始跳转: ${url}`);
      await page.goto(url, { waitUntil: 'domcontentloaded', timeout: 60_000 });
      const data = await this.extractInteractiveContent();
      return toolSuccess(data, '页面导航成功');
    } catch (e) {
      return toolFailure(`页面导航失败: ${(e as Error).message}`);
    }
  }

  async restart(url: string) {
    await this.cleanup();
    await this.initialize();
    return this.navigate(url);
  }

  async click(index?: number, coordinateX?: number, coordinateY?: number) {
    try {
      const page = await this.ensureLatestPage();
      if (index != null) {
        const element = await this.getElementByIndex(index);
        if (!element) return toolFailure('元素不存在');
        const visible = await evaluatePageScriptWithArg<boolean>(page, IS_ELEMENT_VISIBLE_FUNC, element);
        if (!visible) {
          await evaluatePageScriptWithArg<void>(page, SCROLL_ELEMENT_INTO_VIEW_FUNC, element);
          await new Promise((r) => setTimeout(r, 1000));
        }
        await element.click({ timeout: 5000 });
      } else if (coordinateX != null && coordinateY != null) {
        await page.mouse.click(coordinateX, coordinateY);
      }
      return toolSuccess({ url: page.url() }, '点击成功');
    } catch (e) {
      return toolFailure(`点击失败: ${(e as Error).message}`);
    }
  }

  async input(
    text: string,
    pressEnter: boolean,
    index?: number,
    coordinateX?: number,
    coordinateY?: number,
  ) {
    try {
      const page = await this.ensureLatestPage();
      if (index != null) {
        const element = await this.getElementByIndex(index);
        if (!element) return toolFailure('元素不存在');
        try {
          await element.fill('');
          await element.type(text);
        } catch {
          await element.click();
          await element.type(text);
        }
      } else if (coordinateX != null && coordinateY != null) {
        await page.mouse.click(coordinateX, coordinateY);
        await page.keyboard.type(text);
      }
      if (pressEnter) await page.keyboard.press('Enter');
      return toolSuccess({ url: page.url() }, '输入成功');
    } catch (e) {
      return toolFailure(`输入失败: ${(e as Error).message}`);
    }
  }

  async moveMouse(coordinateX?: number, coordinateY?: number) {
    try {
      const page = await this.ensureLatestPage();
      if (coordinateX != null && coordinateY != null) {
        await page.mouse.move(coordinateX, coordinateY);
      }
      return toolSuccess({}, '移动成功');
    } catch (e) {
      return toolFailure(`移动失败: ${(e as Error).message}`);
    }
  }

  async pressKey(key: string) {
    try {
      const page = await this.ensureLatestPage();
      await page.keyboard.press(key);
      return toolSuccess({}, '按键成功');
    } catch (e) {
      return toolFailure(`按键失败: ${(e as Error).message}`);
    }
  }

  async selectOption(index: number, option: number) {
    try {
      const element = await this.getElementByIndex(index);
      if (!element) return toolFailure('元素不存在');
      await element.selectOption({ index: option });
      return toolSuccess({}, '选择成功');
    } catch (e) {
      return toolFailure(`选择失败: ${(e as Error).message}`);
    }
  }

  async scrollUp(toUp?: boolean) {
    const page = await this.ensureLatestPage();
    await evaluatePageScriptWithArg<void>(page, '(toUp) => { if (toUp) window.scrollTo(0, 0); else window.scrollBy(0, -window.innerHeight); }', Boolean(toUp));
    return toolSuccess({}, '滚动成功');
  }

  async scrollDown(toDown?: boolean) {
    const page = await this.ensureLatestPage();
    await evaluatePageScriptWithArg<void>(
      page,
      '(toDown) => { if (toDown) window.scrollTo(0, document.body.scrollHeight); else window.scrollBy(0, window.innerHeight); }',
      Boolean(toDown),
    );
    return toolSuccess({}, '滚动成功');
  }

  async consoleExec(javascript: string) {
    try {
      const page = await this.ensureLatestPage();
      await evaluatePageScript<void>(page, INJECT_CONSOLE_FUNC).catch(() => undefined);
      const result = await page.evaluate(javascript);
      return toolSuccess(result, '执行成功');
    } catch (e) {
      return toolFailure(`执行失败: ${(e as Error).message}`);
    }
  }

  async consoleView(maxLines?: number) {
    try {
      const page = await this.ensureLatestPage();
      const logs = await evaluatePageScriptWithArg<unknown[]>(page, GET_CONSOLE_LOGS_FUNC, maxLines ?? null);
      return toolSuccess(logs, '获取成功');
    } catch (e) {
      return toolFailure(`获取控制台失败: ${(e as Error).message}`);
    }
  }

  async screenshot(fullPage = false): Promise<Buffer> {
    const page = await this.ensureLatestPage();
    return page.screenshot({ fullPage, type: 'png' });
  }
}
